ports/esp32/machine_i2c.c

```diff
@@ -82,126 +66,60 @@ static void machine_hw_i2c_init(machine_hw_i2c_obj_t *self, uint32_t freq, uint3
 }
 
-int machine_hw_i2c_transfer(mp_obj_base_t *self_in, uint16_t addr, size_t n, mp_machine_i2c_buf_t *bufs, unsigned int flags) {
+static int machine_i2c_transfer_single(mp_obj_base_t *self_in, uint16_t addr, size_t len, uint8_t *buf, unsigned int flags) {
     machine_hw_i2c_obj_t *self = MP_OBJ_TO_PTR(self_in);
 
-    /* 0. Probe the address to see if any device responds */
-    esp_err_t err = i2c_master_probe(self->bus_handle, addr, 1000);
+    // 1. Probe the address to see if any device responds.
+    //    This test uses a fixed scl freq of 100_000.
+    esp_err_t err = i2c_master_probe(self->bus_handle, addr, self->timeout_us / 1000);
     if (err != ESP_OK) {
-        return -MP_ENODEV;          /* No device at address, return immediately */
+        return -MP_ENODEV;   // No device at address, return immediately
     }
-    /* 1. Create a temporary device handle for this transaction */
-    i2c_device_config_t dev_cfg = {
-        .dev_addr_length = I2C_ADDR_BIT_LEN_7,
-        .device_address = addr,
-        .scl_speed_hz = 100000,   /* Use bus frequency */
-    };
-    i2c_master_dev_handle_t dev_handle;
-    err = i2c_master_bus_add_device(self->bus_handle, &dev_cfg, &dev_handle);
-    if (err != ESP_OK) {
-        return -MP_ENODEV;
-    }
-
-    int data_len = 0;
 
-    /* 2. If WRITE1 segment exists, perform the write first */
-    if (flags & MP_MACHINE_I2C_FLAG_WRITE1) {
-        if (bufs->len) {
-            err = i2c_master_transmit(dev_handle, bufs->buf, bufs->len, 1000);                 /* Block with 1 s timeout */
-            if (err != ESP_OK) {
-                goto cleanup;
-            }
+    if (len > 0) {
+        // 2. Create a temporary device handle for this transaction
+        //    This step for every transaction can be omitted in
+        //    esp idf v5.5+, which supports handle address changing.
+        i2c_device_config_t dev_cfg = {
+            .dev_addr_length = I2C_ADDR_BIT_LEN_7,
+            .device_address = addr,
+            .scl_speed_hz = self->freq,
+        };
+        i2c_master_dev_handle_t dev_handle;
+        err = i2c_master_bus_add_device(self->bus_handle, &dev_cfg, &dev_handle);
+        if (err != ESP_OK) {
+            return -MP_ENODEV;
         }
-        data_len += bufs->len;
-        --n;
-        ++bufs;
-    }
-    if (flags & MP_MACHINE_I2C_FLAG_READ) {
-        /* 3. Main loop: remaining segments */
-        for (; n--; ++bufs) {
-            if (bufs->len == 0) {
-                continue;
-            }
-            err = i2c_master_receive(dev_handle, bufs->buf, bufs->len, 1000);
-            if (err != ESP_OK) {
-                break;
-            }
-
-            data_len += bufs->len;
+        // 3. Transfer data
+        if (flags & MP_MACHINE_I2C_FLAG_READ) {
+            err = i2c_master_receive(dev_handle, buf, len, self->timeout_us / 1000);
+        } else if (len > 0) {
```

Comment:
This `if (len > 0)` is obsolete. Sorry. It stayed from an earlier version. This line can be just `} else {`.